include/trieste/ast.h

```diff
@@ -155,7 +155,37 @@ namespace trieste
     }
 
   public:
-    ~NodeDef() {}
+    ~NodeDef()
+    {
+      // Contains a pointer to a vector on the stack for handling the
+      // recursive destruction of the AST.
+      // We don't use an actual vector in the TLS as the destruction
+      // of the TLS can cause problems on some platforms.
+      thread_local std::vector<Nodes>* work_list;
+
+      if (work_list)
+      {
+        work_list->push_back(std::move(children));
```

Comment:
Minor suggestion: add a comment here to make it explicit that this will only ever be called by the loop below because the variable is a thread_local. The only circumstance in which this branch is reached is to queue more things to destruct as a result of nodes.clear() on line 184. That was the only stumbling block I had.